src/engine/GroupByImpl.cpp

```diff
@@ -31,6 +31,212 @@
+        }));
+
+      } else if (singleIdTable_) {
+        // Yield at least a single empty table if requested.
+        finalValue_ = LazyResult(lazySingleValueRange([this]() {
+          return IdTableVocabPair{std::move(resultTable_),
+                                  std::move(currentLocalVocab_)};
+        }));
+      } else {
+        // no final values to add
+        isFinished_ = true;
+      }
+      return;
+    }
+
+    // Process remaining items in the last group.  For those we have already
+    // called `lazyGroupBy.processBlock()` but the call to `commitRow` is
+    // still missing. We have to setup a dummy input table and evaluation
+    // context, that have the values of the `currentGroupBlock` in the
+    // correct columns.
+    finalValue_ = LazyResult(lazySingleValueRange([this]() {
+      IdTable idTable{inWidth_, ad_utility::makeAllocatorWithLimit<Id>(
+                                    1_B * sizeof(Id) * inWidth_)};
+      idTable.emplace_back();
+      for (const auto& [colIdx, value] : currentGroupBlock_) {
+        idTable.at(0, colIdx) = value;
+      }
+
+      sparqlExpression::EvaluationContext evaluationContext =
+          parent_->createEvaluationContext(currentLocalVocab_, idTable);
+
+      lazyGroupBy_->commitRow(resultTable_, evaluationContext,
+                              currentGroupBlock_);
+      currentLocalVocab_.mergeWith(storedLocalVocabs_);
+      return IdTableVocabPair{std::move(resultTable_),
+                              std::move(currentLocalVocab_)};
+    }));
+  }
+
+  void yieldFinalValue() {
+    buildFinalValue();
+    if (!finalValue_.has_value()) {
+      isFinished_ = true;
+      return;
+    }
+
+    // calling begin ensures that the final value is ready
+    it_ = ql::ranges::begin(finalValue_.value());
+  }
+
+  void start() {
+    using namespace ad_utility;
+    lazyGroupBy_ = std::make_unique<LazyGroupBy>(
+        currentLocalVocab_, std::move(aggregateAliases_),
+        parent_->getExecutionContext()->getAllocator(), groupByCols_.size());
+
+    range_ = Result::LazyResult(CachingContinuableTransformInputRange(
+        subresult_->idTables(),
+        [this](Result::IdTableVocabPair& pair) { return process(pair); }));
+    it_ = ql::ranges::begin(range_.value());
+    // if the main range is empty, yield the final value.
+    if (it_ == ql::ranges::end(range_.value())) {
+      yieldFinalValue();
+    }
+  }
+
+  void next() {
+    // since  the final value is a single-value range, the value has already
+    // been yielded if the range exists
+    if (finalValue_.has_value()) {
+      AD_CORRECTNESS_CHECK(++it_ == ql::ranges::end(finalValue_.value()));
+      isFinished_ = true;
+      return;
+    }
+
+    it_++;
+    if (it_ == ql::ranges::end(range_.value())) {
+      yieldFinalValue();
+    }
+  }
+
+  bool isFinished() { return isFinished_; }
+  auto& get() { return *it_; }
+  const auto& get() const { return *it_; }
+};
```

Comment:
In general I like this class (the confusion comes from the original code)

@RobinTF Can you please also have a look at this (this was originally your code, and all the lifetime and corner cases are not inside my expertise).